src/net/socket_address.rs

```diff
@@ -0,0 +1,72 @@
+#![allow(unsafe_code)]
+use core::mem::{self, size_of};
+
+use crate::backend::c;
+use crate::net::SocketAddr;
+
+use super::SocketAddrStorage;
+
+/// A trait abstracting over the types that can be passed as a `sockaddr`.
+///
+/// Safety: by implementing this trait, you assert that the values returned
+/// by the trait methods can be passed to the system calls that accept `sockaddr`.
+pub unsafe trait SocketAddress {
+    /// The corresponding C `sockaddr_*` type.
+    type CSockAddr;
+
+    /// Convert to the C type.
+    fn encode(&self) -> Self::CSockAddr;
```

Comment:
I find it a bit annoying/confusing that this trait uses default fns. We could instead utilize a small custom enum that stores either a reference to or an owned `CSockAddr`. Kinda like a `Cow` but for nostd, something along the lines of

```rust
/// Borrowed Or oWned
pub enum Bow<'a, B> {
    /// Stores a reference to the value.
    Borrowed(&'a B),
    /// Owns the value.
    Owned(B),
}

impl<'a, B> Bow<'a, B> {
    /// Get a reference to the value stored in this Bow.
    pub fn as_ref(&self) -> &B {
        match self {
            Bow::Borrowed(val) => val,
            Bow::Owned(val) => &val,
        }
    }
}
```

Then only `encode` would be enough and the rest could be defined as helper functions or similar.

Let me know what you think

Reply:
I removed the `write_sockaddr` method, which had a default implementation that was not overridden for any of the types. It was only used in one place, for `SocketAddrAny::write`.

The purpose of the `with_sockaddr` method is to avoid the copying that you mention, and is overridden in the implementation for `SocketAddrUnix` which contains a `sockaddr_un`.

Before simplifying this trait further, I'd like to hear what @sunfishcode thinks about the `CSockAddr` associated type. If it's [undesirable to expose the underlying C types](https://github.com/bytecodealliance/rustix/pull/956), I think the way to go would be to remove `CSockAddr` and `encode`, leaving only `with_sockaddr`. This is a messier interface than simply returning a value, but a private helper function could be used to implement this in terms of the `encode_sockaddr_*` functions for the majority of the address types.